File: javascript/hanoi_tower.js
let moveCount = 0;

function hanoiTower(n, from, to, aux) {
    if (n === 1) {
        moveCount++;
        return;
    }
    
    hanoiTower(n - 1, from, aux, to);
    moveCount++;
    hanoiTower(n - 1, aux, to, from);
}

// 兼容 Node.js 和 QuickJS
const args = typeof process !== 'undefined' ? process.argv.slice(2) : scriptArgs.slice(1);
if (args.length < 1) {
    console.log('Usage: hanoi_tower <n>');
    if (typeof process !== 'undefined') process.exit(1);
    else throw new Error('Usage error');
}

const n = parseInt(args[0]);
if (isNaN(n)) {
    console.log('Invalid number:', args[0]);
    if (typeof process !== 'undefined') process.exit(1);
    else throw new Error('Usage error');
}

moveCount = 0;

const start = Date.now();
hanoiTower(n, 'A', 'C', 'B');
const end = Date.now();

const timeMs = end - start;

console.log(`JavaScript: hanoi_tower(${n}) = ${moveCount}`);
console.log(`Time: ${timeMs}ms`);

File: python/hanoi_tower.py
import sys
import time

move_count = 0

def hanoi_tower(n, from_pole, to_pole, aux_pole):
    global move_count
    if n == 1:
        move_count += 1
        return
    
    hanoi_tower(n - 1, from_pole, aux_pole, to_pole)
    move_count += 1
    hanoi_tower(n - 1, aux_pole, to_pole, from_pole)

if len(sys.argv) < 2:
    print('Usage: hanoi_tower <n>')
    sys.exit(1)

try:
    n = int(sys.argv[1])
except ValueError:
    print('Invalid number:', sys.argv[1])
    sys.exit(1)

move_count = 0

start = time.time()
hanoi_tower(n, 'A', 'C', 'B')
end = time.time()

time_ms = int((end - start) * 1000)

print(f'Python: hanoi_tower({n}) = {move_count}')
print(f'Time: {time_ms}ms')